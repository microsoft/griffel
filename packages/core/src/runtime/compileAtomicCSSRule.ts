import { hyphenateProperty } from './utils/hyphenateProperty';
import { normalizeNestedProperty } from './utils/normalizeNestedProperty';
import { compileCSSRules } from './compileCSSRules';

export interface CompileAtomicCSSOptions {
  className: string;

  selectors: string[];
  media: string;
  layer: string;
  support: string;
  container: string;

  property: string;
  value: number | string | Array<number | string>;

  rtlClassName?: string;
  rtlProperty?: string;
  rtlValue?: number | string | Array<number | string>;
}

const PSEUDO_SELECTOR_REGEX = /,( *[^ &])/g;

/**
 * Normalizes pseudo selectors to always contain &, requires to work properly with comma-separated selectors.
 *
 * @example
 *   ":hover" => "&:hover"
 *   " :hover" => "& :hover"
 *   ":hover,:focus" => "&:hover,&:focus"
 *   " :hover, :focus" => "& :hover,& :focus"
 */
export function normalizePseudoSelector(pseudoSelector: string): string {
  return (
    '&' +
    normalizeNestedProperty(
      // Regex there replaces a comma, spaces and an ampersand if it's present with comma and an ampersand.
      // This allows to normalize input, see examples in JSDoc.
      pseudoSelector.replace(PSEUDO_SELECTOR_REGEX, ',&$1'),
    )
  );
}

function createCSSRule(classNameSelector: string, cssDeclaration: string, pseudos: string[]): string {
  let cssRule = cssDeclaration;

  if (pseudos.length > 0) {
    cssRule = pseudos.reduceRight((acc, selector) => {
      return `${normalizePseudoSelector(selector)} { ${acc} }`;
    }, cssDeclaration);
  }

  return `${classNameSelector}{${cssRule}}`;
}

export function compileAtomicCSSRule(
  options: CompileAtomicCSSOptions,
): [string? /* ltr definition */, string? /* rtl definition */] {
  const {
    className,
    media,
    layer,
    selectors,
    support,
    property,
    rtlClassName,
    rtlProperty,
    rtlValue,
    value,
    container,
  } = options;

  const classNameSelector = `.${className}`;
  const cssDeclaration = Array.isArray(value)
    ? `${value.map(v => `${hyphenateProperty(property)}: ${v}`).join(';')};`
    : `${hyphenateProperty(property)}: ${value};`;

  let cssRule = createCSSRule(classNameSelector, cssDeclaration, selectors);

  if (rtlProperty && rtlClassName) {
    const rtlClassNameSelector = `.${rtlClassName}`;
    const rtlCSSDeclaration = Array.isArray(rtlValue)
      ? `${rtlValue.map(v => `${hyphenateProperty(rtlProperty)}: ${v}`).join(';')};`
      : `${hyphenateProperty(rtlProperty)}: ${rtlValue};`;

    cssRule += createCSSRule(rtlClassNameSelector, rtlCSSDeclaration, selectors);
  }

  if (media) {
    cssRule = `@media ${media} { ${cssRule} }`;
  }

  if (layer) {
    cssRule = `@layer ${layer} { ${cssRule} }`;
  }

  if (support) {
    cssRule = `@supports ${support} { ${cssRule} }`;
  }

  if (container) {
    cssRule = `@container ${container} { ${cssRule} }`;
  }

  return compileCSSRules(cssRule, true) as [string?, string?];
}
