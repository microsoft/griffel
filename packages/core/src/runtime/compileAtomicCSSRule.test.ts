import { compileAtomicCSSRule, normalizePseudoSelector } from './compileAtomicCSSRule';
import type { CompileAtomicCSSOptions } from './compileAtomicCSSRule';

const defaultOptions: Pick<
  CompileAtomicCSSOptions,
  'rtlClassName' | 'className' | 'media' | 'selectors' | 'support' | 'layer' | 'container'
> = {
  className: 'foo',
  rtlClassName: 'rtl-foo',
  media: '',
  selectors: [],
  support: '',
  layer: '',
  container: '',
};

describe('compileAtomicCSSRule', () => {
  it.each([
    ['transform', 'none'],
    ['flex-grow', '1'],
    ['flex-shrink', '1'],
    ['flex-basis', 'auto'],
    ['flex-direction', 'column'],
    ['align-self', 'center'],
    ['align-content', 'center'],
    ['align-items', 'center'],
    ['order', '1'],
    ['justify-content', 'center'],
    ['display', 'flex'],
    ['display', 'inline-flex'],
    ['display', 'grid'],
    ['display', 'inline-grid'],
    ['transition', 'margin-right 4s'],
    ['writing-mode', 'vertical-lr'],
    ['columns', '2'],
    ['text-size-adjust', 'none'],
    ['text-decoration', 'none'],
    ['filter', 'blur(5px)'],
    ['position', 'sticky'],
    ['clip-path', 'circle(40%)'],
    ['width', 'fit-content'],
    ['width', 'min-block-size'],
    ['background-clip', 'content-box'],
    ['animation', '3s linear 1s slidein;'],
    ['animation-delay', '3s'],
    ['animation-direction', 'normal'],
    ['animation-duration', '3s'],
    ['animation-fill-mode', 'both'],
    ['animation-iteration-count', '10'],
    ['animation-name', 'foo'],
    ['animation-play-state', 'running'],
    ['animation-timing-function', 'linear'],
    ['scroll-snap-type', 'none'],
    ['scroll-margin', '0'],
  ])('does not prefix %s:%s', (property, value) => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        property,
        value,
      }),
    ).toMatchSnapshot();
  });

  it('handles pseudo', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        selectors: [':hover'],
        property: 'color',
        value: 'red',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo:hover{color:red;}",
      ]
    `);
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        selectors: [':focus:hover'],
        property: 'color',
        value: 'red',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo:focus:hover{color:red;}",
      ]
    `);
  });

  it('handles array of values', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        property: 'color',
        value: ['red', 'blue'],
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo{color:red;color:blue;}",
      ]
    `);
  });

  it('handles at-rules', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        media: '(max-width: 100px)',
        property: 'color',
        value: 'red',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        "@media (max-width: 100px){.foo{color:red;}}",
      ]
    `);
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        support: '(display: table-cell)',
        property: 'color',
        value: 'red',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        "@supports (display: table-cell){.foo{color:red;}}",
      ]
    `);
  });

  it('handles rtl properties', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,

        property: 'paddingLeft',
        value: '10px',

        rtlProperty: 'paddingRight',
        rtlValue: '10px',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo{padding-left:10px;}",
        ".rtl-foo{padding-right:10px;}",
      ]
    `);
  });

  it('handles rtl properties with pseudo selectors', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        selectors: [':before'],

        property: 'paddingLeft',
        value: '10px',

        rtlProperty: 'paddingRight',
        rtlValue: '10px',
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo:before{padding-left:10px;}",
        ".rtl-foo:before{padding-right:10px;}",
      ]
    `);
  });

  it('handles rtl properties with fallback values', () => {
    expect(
      compileAtomicCSSRule({
        ...defaultOptions,
        property: 'paddingLeft',
        value: [0, '10px'],
        rtlProperty: 'paddingRight',
        rtlValue: [0, '10px'],
      }),
    ).toMatchInlineSnapshot(`
      Array [
        ".foo{padding-left:0;padding-left:10px;}",
        ".rtl-foo{padding-right:0;padding-right:10px;}",
      ]
    `);
  });

  describe('global', () => {
    it('compiles global rules', () => {
      expect(
        compileAtomicCSSRule({
          ...defaultOptions,
          selectors: [':global(body)'],
          property: 'color',
          value: 'red',
        }),
      ).toMatchInlineSnapshot(`
              Array [
                "body .foo{color:red;}",
              ]
          `);
      expect(
        compileAtomicCSSRule({
          ...defaultOptions,
          selectors: [':global(.fui-FluentProvider)', '& .focus:hover'],
          property: 'color',
          value: 'red',
        }),
      ).toMatchInlineSnapshot(`
              Array [
                ".fui-FluentProvider .foo .focus:hover{color:red;}",
              ]
          `);
    });

    it('compiles global rules with RTL', () => {
      expect(
        compileAtomicCSSRule({
          ...defaultOptions,
          selectors: [':global(body)'],
          property: 'paddingLeft',
          value: '10px',

          rtlProperty: 'paddingRight',
          rtlValue: '10px',
        }),
      ).toMatchInlineSnapshot(`
              Array [
                "body .foo{padding-left:10px;}",
                "body .rtl-foo{padding-right:10px;}",
              ]
          `);
    });
  });
});

describe('normalizePseudoSelector', () => {
  it('handles basic ', () => {
    expect(normalizePseudoSelector(':hover')).toMatchInlineSnapshot(`"&:hover"`);
  });

  it('handles spacing', () => {
    expect(normalizePseudoSelector(' :hover')).toMatchInlineSnapshot(`"& :hover"`);
  });

  it('handles multiple pseudos', () => {
    expect(normalizePseudoSelector(':focus:hover')).toMatchInlineSnapshot(`"&:focus:hover"`);
  });

  it('handles comma separated pseudos', () => {
    expect(normalizePseudoSelector('& :hover, & :focus')).toMatchInlineSnapshot(`"& :hover, & :focus"`);
    expect(normalizePseudoSelector(':focus,:hover')).toMatchInlineSnapshot(`"&:focus,&:hover"`);
    expect(normalizePseudoSelector(':focus, :hover')).toMatchInlineSnapshot(`"&:focus,& :hover"`);
  });
});
