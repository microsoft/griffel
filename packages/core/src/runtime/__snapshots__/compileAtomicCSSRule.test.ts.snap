// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileAtomicCSSRule does not prefix align-content:center 1`] = `
Array [
  ".foo{align-content:center;}",
]
`;

exports[`compileAtomicCSSRule does not prefix align-items:center 1`] = `
Array [
  ".foo{align-items:center;}",
]
`;

exports[`compileAtomicCSSRule does not prefix align-self:center 1`] = `
Array [
  ".foo{align-self:center;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation:3s linear 1s slidein; 1`] = `
Array [
  ".foo{animation:3s linear 1s slidein;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-delay:3s 1`] = `
Array [
  ".foo{animation-delay:3s;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-direction:normal 1`] = `
Array [
  ".foo{animation-direction:normal;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-duration:3s 1`] = `
Array [
  ".foo{animation-duration:3s;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-fill-mode:both 1`] = `
Array [
  ".foo{animation-fill-mode:both;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-iteration-count:10 1`] = `
Array [
  ".foo{animation-iteration-count:10;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-name:foo 1`] = `
Array [
  ".foo{animation-name:foo;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-play-state:running 1`] = `
Array [
  ".foo{animation-play-state:running;}",
]
`;

exports[`compileAtomicCSSRule does not prefix animation-timing-function:linear 1`] = `
Array [
  ".foo{animation-timing-function:linear;}",
]
`;

exports[`compileAtomicCSSRule does not prefix background-clip:content-box 1`] = `
Array [
  ".foo{background-clip:content-box;}",
]
`;

exports[`compileAtomicCSSRule does not prefix clip-path:circle(40%) 1`] = `
Array [
  ".foo{clip-path:circle(40%);}",
]
`;

exports[`compileAtomicCSSRule does not prefix columns:2 1`] = `
Array [
  ".foo{columns:2;}",
]
`;

exports[`compileAtomicCSSRule does not prefix display:flex 1`] = `
Array [
  ".foo{display:flex;}",
]
`;

exports[`compileAtomicCSSRule does not prefix display:grid 1`] = `
Array [
  ".foo{display:grid;}",
]
`;

exports[`compileAtomicCSSRule does not prefix display:inline-flex 1`] = `
Array [
  ".foo{display:inline-flex;}",
]
`;

exports[`compileAtomicCSSRule does not prefix display:inline-grid 1`] = `
Array [
  ".foo{display:inline-grid;}",
]
`;

exports[`compileAtomicCSSRule does not prefix filter:blur(5px) 1`] = `
Array [
  ".foo{filter:blur(5px);}",
]
`;

exports[`compileAtomicCSSRule does not prefix flex-basis:auto 1`] = `
Array [
  ".foo{flex-basis:auto;}",
]
`;

exports[`compileAtomicCSSRule does not prefix flex-direction:column 1`] = `
Array [
  ".foo{flex-direction:column;}",
]
`;

exports[`compileAtomicCSSRule does not prefix flex-grow:1 1`] = `
Array [
  ".foo{flex-grow:1;}",
]
`;

exports[`compileAtomicCSSRule does not prefix flex-shrink:1 1`] = `
Array [
  ".foo{flex-shrink:1;}",
]
`;

exports[`compileAtomicCSSRule does not prefix justify-content:center 1`] = `
Array [
  ".foo{justify-content:center;}",
]
`;

exports[`compileAtomicCSSRule does not prefix order:1 1`] = `
Array [
  ".foo{order:1;}",
]
`;

exports[`compileAtomicCSSRule does not prefix position:sticky 1`] = `
Array [
  ".foo{position:sticky;}",
]
`;

exports[`compileAtomicCSSRule does not prefix scroll-margin:0 1`] = `
Array [
  ".foo{scroll-margin:0;}",
]
`;

exports[`compileAtomicCSSRule does not prefix scroll-snap-type:none 1`] = `
Array [
  ".foo{scroll-snap-type:none;}",
]
`;

exports[`compileAtomicCSSRule does not prefix text-decoration:none 1`] = `
Array [
  ".foo{text-decoration:none;}",
]
`;

exports[`compileAtomicCSSRule does not prefix text-size-adjust:none 1`] = `
Array [
  ".foo{text-size-adjust:none;}",
]
`;

exports[`compileAtomicCSSRule does not prefix transform:none 1`] = `
Array [
  ".foo{transform:none;}",
]
`;

exports[`compileAtomicCSSRule does not prefix transition:margin-right 4s 1`] = `
Array [
  ".foo{transition:margin-right 4s;}",
]
`;

exports[`compileAtomicCSSRule does not prefix width:fit-content 1`] = `
Array [
  ".foo{width:fit-content;}",
]
`;

exports[`compileAtomicCSSRule does not prefix width:min-block-size 1`] = `
Array [
  ".foo{width:min-block-size;}",
]
`;

exports[`compileAtomicCSSRule does not prefix writing-mode:vertical-lr 1`] = `
Array [
  ".foo{writing-mode:vertical-lr;}",
]
`;
